lib/spack/spack/spec.py

```diff
@@ -2436,6 +2436,9 @@ def ensure_no_deprecated(root):
             raise SpecDeprecatedError(msg)
 
     def _new_concretize(self, tests=False):
+        if self._concrete:
+            return
+
```

Comment:
This is already in, see #20196 . Also: local import statements should stay at the top to avoid errors where we use something that is being defined later.